pokemon.go

```diff
@@ -230,39 +230,115 @@ func (p *Pokemon) computeStats() {
 	if !p.HasValidEVs() {
 		panic(fmt.Sprintf("Failed to compute stats for %s, evs %d invalid", p.GetName(), p.EVs))
 	}
+	// Set stats given base values
+	baseStats := p.GetBaseStats()
+	p.updateMaxHP(baseStats[StatHP])
+	for _, stat := range []int{StatAtk, StatDef, StatSpAtk, StatSpDef, StatSpeed} {
+		base := float64(baseStats[stat])
+		// Other stat formula
+		ev_term := math.Floor(float64(p.EVs[stat]) / 4)
+		base_iv_ev_level_term := (2*uint(base) + uint(p.IVs[stat]) + uint(ev_term)) * uint(p.Level)
+		base = math.Floor(float64(base_iv_ev_level_term)/100 + 5)
+		n := p.Nature
+		if n.StatUp != n.StatDown {
+			if n.StatUp == stat {
+				base *= 1.1
+			}
+			if n.StatDown == stat {
+				base *= 0.9
+			}
+		}
+		switch stat {
+		case StatAtk:
+			p.updateAttack(base)
+		case StatDef:
+			p.updateDefense(base)
+		case StatSpAtk:
+			p.updateSpecialAttack(base)
+		case StatSpDef:
+			p.updateSpecialDefense(base)
+		case StatSpeed:
+			p.updateSpeed(base)
+		}
+	}
+}
 
-	// get base stats
-	base_stats := p.GetBaseStats()
-
-	// compute HP
+// Stat setters (helpers for computeStats)
+func (p *Pokemon) updateMaxHP(base int) {
+	// Base calculation
 	hp_ev_term := math.Floor(float64(p.EVs[StatHP]) / 4)
-	base_iv_ev_level_hp_term := (2*uint(base_stats[StatHP]) + uint(p.IVs[StatHP]) + uint(hp_ev_term)) * uint(p.Level)
+	base_iv_ev_level_hp_term := (2*uint(base) + uint(p.IVs[StatHP]) + uint(hp_ev_term)) * uint(p.Level)
 	hp_floor_term := math.Floor(float64(base_iv_ev_level_hp_term) / 100)
-	p.Stats[StatHP] = uint(hp_floor_term + float64(p.Level) + 10)
-	p.CurrentHP = p.Stats[StatHP]
+	effective := uint(hp_floor_term + float64(p.Level) + 10)
+	// TODO: Max HP modifiers
+	p.Stats[StatHP] = effective
+	p.CurrentHP = effective
+}
 
-	// compute all other stats
-	natureModifiers := p.Nature.getNatureModifers()
-	for _, stat := range [5]int{StatAtk, StatDef, StatSpAtk, StatSpDef, StatSpeed} {
-		ev_term := math.Floor(float64(p.EVs[stat]) / 4)
-		base_iv_ev_level_term := (2*uint(base_stats[stat]) + uint(p.IVs[stat]) + uint(ev_term)) * uint(p.Level)
-		floor_term := math.Floor(float64(base_iv_ev_level_term) / 100)
-		nature_term := float64(floor_term+5) * natureModifiers[stat]
-		p.Stats[stat] = uint(math.Floor(nature_term))
-	}
+func (p *Pokemon) updateAttack(base float64) {
+	effective := base
+	// TODO: attack modifiers
+	p.Stats[StatAtk] = uint(effective)
```

Comment:
my instinctual thought is to be wary of overwriting the stats field with effective stats. this is because I'm afraid that we'll encounter a move or ability whose effect scales with a Pokemon's base stats rather than their effective stats, though I cannot think of any off of the top of my head.

what if we stored the effective stats in another field?

Reply:
or we could use a method to compute the effective stats and return an effective stat array

Reply:
If we need the base stats, can't we just grab them again? Shouldn't `Stats` _be_ the effective stats though?

Reply:
Agree here that stats is the effective stats. This PR is concerned with the way the effective stats are calculated/updated. Base stats are always available from `p.GetBaseStats()`